File: src/discovery/OSDetector.js
'use strict';

/**
 * Detects the operating system and distribution of a remote server.
 */
class OSDetector {
  constructor(sshManager) {
    this.ssh = sshManager;
  }

  /**
   * Detect the full OS information from the remote server.
   * @returns {Promise<object>} OS details including type, distro, version, kernel, arch.
   */
  async detect() {
    const results = await Promise.allSettled([
      this.ssh.execute('uname -a'),
      this.ssh.execute('cat /etc/os-release 2>/dev/null || cat /etc/redhat-release 2>/dev/null || cat /etc/issue 2>/dev/null'),
      this.ssh.execute('uname -m'),
      this.ssh.execute('hostname'),
      this.ssh.execute('uptime -p 2>/dev/null || uptime'),
      this.ssh.execute('cat /proc/version 2>/dev/null'),
      this.ssh.execute('lsb_release -a 2>/dev/null'),
    ]);

    const uname = results[0].status === 'fulfilled' ? results[0].value.stdout : '';
    const osRelease = results[1].status === 'fulfilled' ? results[1].value.stdout : '';
    const arch = results[2].status === 'fulfilled' ? results[2].value.stdout : '';
    const hostname = results[3].status === 'fulfilled' ? results[3].value.stdout : '';
    const uptime = results[4].status === 'fulfilled' ? results[4].value.stdout : '';
    const procVersion = results[5].status === 'fulfilled' ? results[5].value.stdout : '';
    const lsbRelease = results[6].status === 'fulfilled' ? results[6].value.stdout : '';

    const osInfo = {
      type: this._detectOSType(uname),
      distribution: this._parseDistribution(osRelease, lsbRelease),
      version: this._parseVersion(osRelease, lsbRelease),
      kernel: this._parseKernel(uname, procVersion),
      architecture: arch,
      hostname: hostname,
      uptime: uptime,
      raw: { uname, osRelease, procVersion, lsbRelease },
    };

    return osInfo;
  }

  _detectOSType(uname) {
    const lower = uname.toLowerCase();
    if (lower.includes('linux')) return 'Linux';
    if (lower.includes('darwin')) return 'macOS';
    if (lower.includes('freebsd')) return 'FreeBSD';
    if (lower.includes('openbsd')) return 'OpenBSD';
    if (lower.includes('cygwin') || lower.includes('mingw')) return 'Windows (Cygwin)';
    return 'Unknown';
  }

  _parseDistribution(osRelease, lsbRelease) {
    // Try os-release first (most modern distros)
    const nameMatch = osRelease.match(/^NAME="?([^"\n]+)"?/m);
    if (nameMatch) return nameMatch[1].trim();

    // Try lsb_release
    const lsbMatch = lsbRelease.match(/Distributor ID:\s*(.+)/);
    if (lsbMatch) return lsbMatch[1].trim();

    // Try common patterns
    if (osRelease.includes('CentOS')) return 'CentOS';
    if (osRelease.includes('Red Hat')) return 'Red Hat Enterprise Linux';
    if (osRelease.includes('Ubuntu')) return 'Ubuntu';
    if (osRelease.includes('Debian')) return 'Debian';
    if (osRelease.includes('Amazon Linux')) return 'Amazon Linux';
    if (osRelease.includes('Alpine')) return 'Alpine Linux';

    return 'Unknown';
  }

  _parseVersion(osRelease, lsbRelease) {
    const versionMatch = osRelease.match(/^VERSION_ID="?([^"\n]+)"?/m);
    if (versionMatch) return versionMatch[1].trim();

    const lsbMatch = lsbRelease.match(/Release:\s*(.+)/);
    if (lsbMatch) return lsbMatch[1].trim();

    return 'Unknown';
  }

  _parseKernel(uname, procVersion) {
    const parts = uname.split(' ');
    if (parts.length >= 3) return parts[2];

    const match = procVersion.match(/version\s+(\S+)/);
    if (match) return match[1];

    return 'Unknown';
  }
}

module.exports = OSDetector;

File: sysadmin_agent/discovery/os_detector.py
from concurrent.futures import ThreadPoolExecutor, as_completed
import re


class OSDetector:
    def __init__(self, ssh_manager):
        self.ssh = ssh_manager

    def detect(self):
        commands = {
            "uname": "uname -a",
            "os_release": "cat /etc/os-release 2>/dev/null || true",
            "arch": "uname -m",
            "hostname": "hostname",
            "uptime": "uptime",
            "proc_version": "cat /proc/version 2>/dev/null || true",
            "lsb_release": "lsb_release -a 2>/dev/null || true",
        }

        results = {}
        with ThreadPoolExecutor(max_workers=3) as executor:
            futures = {
                executor.submit(self.ssh.execute, cmd): name
                for name, cmd in commands.items()
            }
            for future in as_completed(futures):
                name = futures[future]
                try:
                    results[name] = future.result()
                except Exception:
                    results[name] = {"stdout": "", "stderr": "", "exit_code": 1}

        return {
            "type": self._parse_os_type(results),
            "distribution": self._parse_distribution(results),
            "version": self._parse_version(results),
            "kernel": self._parse_kernel(results),
            "architecture": self._parse_architecture(results),
            "hostname": self._parse_hostname(results),
            "uptime": self._parse_uptime(results),
            "raw": {k: v["stdout"] for k, v in results.items()},
        }

    def _parse_os_type(self, results):
        uname = results.get("uname", {}).get("stdout", "").lower()
        if "linux" in uname:
            return "Linux"
        if "darwin" in uname:
            return "macOS"
        if "freebsd" in uname:
            return "FreeBSD"
        return "Unknown"

    def _parse_distribution(self, results):
        os_release = results.get("os_release", {}).get("stdout", "")
        lsb = results.get("lsb_release", {}).get("stdout", "")

        id_match = re.search(r'^ID=[""]?([^""\n]+)', os_release, re.MULTILINE)
        if id_match:
            distro_id = id_match.group(1).strip().lower()
            distro_map = {
                "ubuntu": "Ubuntu",
                "centos": "CentOS",
                "debian": "Debian",
                "almalinux": "AlmaLinux",
                "rocky": "Rocky",
                "amzn": "Amazon Linux",
                "alpine": "Alpine",
                "fedora": "Fedora",
                "rhel": "RHEL",
                "opensuse-leap": "openSUSE Leap",
                "opensuse-tumbleweed": "openSUSE Tumbleweed",
                "arch": "Arch Linux",
                "ol": "Oracle Linux",
            }
            for key, name in distro_map.items():
                if distro_id == key:
                    return name

            name_match = re.search(r'^NAME=[""]?([^""\n]+)', os_release, re.MULTILINE)
            if name_match:
                return name_match.group(1).strip()
            return distro_id.capitalize()

        dist_match = re.search(r'Distributor ID:\s*(.+)', lsb)
        if dist_match:
            return dist_match.group(1).strip()

        uname = results.get("uname", {}).get("stdout", "").lower()
        if "darwin" in uname:
            return "macOS"
        if "freebsd" in uname:
            return "FreeBSD"

        return "Unknown"

    def _parse_version(self, results):
        os_release = results.get("os_release", {}).get("stdout", "")
        match = re.search(r'^VERSION_ID=[""]?([^""\n]+)', os_release, re.MULTILINE)
        if match:
            return match.group(1).strip()

        lsb = results.get("lsb_release", {}).get("stdout", "")
        match = re.search(r'Release:\s*(.+)', lsb)
        if match:
            return match.group(1).strip()

        return "Unknown"

    def _parse_kernel(self, results):
        uname = results.get("uname", {}).get("stdout", "").strip()
        parts = uname.split()
        if len(parts) >= 3:
            return parts[2]
        return uname or "Unknown"

    def _parse_architecture(self, results):
        arch = results.get("arch", {}).get("stdout", "").strip()
        return arch or "Unknown"

    def _parse_hostname(self, results):
        hostname = results.get("hostname", {}).get("stdout", "").strip()
        return hostname or "Unknown"

    def _parse_uptime(self, results):
        uptime = results.get("uptime", {}).get("stdout", "").strip()
        return uptime or "Unknown"
